src/Compilers/CSharp/Portable/Symbols/ConstraintsHelper.cs

```diff
@@ -390,6 +390,55 @@ private static bool CheckConstraintsSingleType(TypeSymbol type, CheckConstraints
         }
 
         public static bool CheckConstraints(
+            this TupleTypeSymbol tuple,
+            ConversionsBase conversions,
+            SyntaxNode typeSyntax,
+            ImmutableArray<Location> elementLocations,
+            Compilation currentCompilation,
+            DiagnosticBag diagnostics)
+        {
+            NamedTypeSymbol type = tuple.TupleUnderlyingType;
+            if (!RequiresChecking(type))
+            {
+                return true;
+            }
+
+            bool result = !typeSyntax.HasErrors;
+
+            var diagnosticsBuilder = ArrayBuilder<TypeParameterDiagnosticInfo>.GetInstance();
+            var underlyingTupleTypeChain = ArrayBuilder<NamedTypeSymbol>.GetInstance();
+            TupleTypeSymbol.GetUnderlyingTypeChain(type, underlyingTupleTypeChain);
+
+            int offset = 0;
+            foreach (var underlyingTuple in underlyingTupleTypeChain)
+            {
+                ArrayBuilder<TypeParameterDiagnosticInfo> useSiteDiagnosticsBuilder = null;
+                result &= CheckTypeConstraints(underlyingTuple, conversions, currentCompilation, diagnosticsBuilder, ref useSiteDiagnosticsBuilder);
+
+                if (useSiteDiagnosticsBuilder != null)
+                {
+                    diagnosticsBuilder.AddRange(useSiteDiagnosticsBuilder);
+                }
+
+                foreach (var pair in diagnosticsBuilder)
+                {
+                    var ordinal = pair.TypeParameter.Ordinal + offset;
```

Comment:
In case of an 8-element tuple, the 8th item in the "rest" part is itself a one-element tuple. Therefor eGetUnderlyingTypeChain returns two ValueTuple. In the second iteration offset is 7, and TypeParameter is zero — the diagnostic will be reported at the 7th position (0-based). This is covered in LongTupleConstraints test.